### PR TITLE
[processor/metricsgeneration] Generate metrics even when the second metric's value is 0

### LINKED ISSUE #35533: [processor/metricsgeneration] Metrics should be calculated when second metric value is 0
### Component(s)

processor/metricsgeneration

### Describe the issue you're reporting

**Bug**
The metrics generation processor is used to create new metrics by using calculations on existing metrics. As explained in the configuration example [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricsgenerationprocessor#configuration), the config specifies the first metric, second metric, and operation type for calculations.

Existing logic for this processor simply skips the calculation when the value of the second metric's datapoint is 0 ([reference](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5f96c1ae634456a382a64caaa61b1a7839a54765/processor/metricsgenerationprocessor/processor.go#L64)). This means that even when it's a valid operation, and the result may be relevant for users, no metric is created.

**Example showing bug**
Relevant example configuration:
```
experimental_metricsgeneration:
  rules:
    - name: available_capacity
      unit: bytes
      type: calculate
      metric1: total_capacity
      metric2: used_capacity
      operation: subtract
```

This configuration should create a new metric called `available_capacity`, which will be the result of `total_capacity - used_capacity`. If the value of `used_capacity` is 0, the processor won't create a new metric called `available_capacity`, when I believe it should.

**Proposed Solution**
I think the solution here is to allow the calculation to be done and to create the resulting metric.

**Open Question**
How should we handle the `divide` and `percent` operations for the case of dividing by 0? We can either add a data point with a value of `0` or not add any data point at all. Since dividing by 0 is an invalid operation, I believe the right thing to do would be log an error and not add a data point in that case. I'm open to feedback from others though.

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

Originally, this processor would not generate a new metric if the value of the metric configured as `metric2` was `0`. There are valid reasons why this calculation should still be done, and a new metric should be calculated.

**Link to tracking Issue:** <Issue number if applicable>
Resolves #35533

**Testing:** <Describe what testing was performed and which tests were added.>
Added tests for each calculation type to show expected results when the second metric's value is 0.

### CODE REVIEW

File: .chloggen/metricsgeneration_metric2_0.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
```
Comment (crobert-1) [inline .chloggen/metricsgeneration_metric2_0.yaml:4]: I consider this to be a breaking change as existing collector configurations will now potentially be emitting more metrics than before.

File: processor/metricsgenerationprocessor/testdata/input_metric_types/config.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+experimental_metricsgeneration/sum_gauge_metric:
+  rules:
+    - name: system.filesystem.capacity
+      unit: bytes
+      type: calculate
+      metric1: system.filesystem.usage
+      metric2: system.filesystem.utilization
+      operation: divide
```
Comment (crobert-1) [inline processor/metricsgenerationprocessor/testdata/input_metric_types/config.yaml:8]:
```suggestion
      operation: divide

```

File: processor/metricsgenerationprocessor/processor_test.go [hunk 3]
```diff
@@ -391,76 +391,101 @@ func getOutputForIntGaugeTest() pmetric.Metrics {
 	return intGaugeOutputMetrics
 }
 
-func TestSumCalculateNewMetric(t *testing.T) {
-	next := new(consumertest.MetricsSink)
-	cfg := &Config{
```
Comment (crobert-1) [inline processor/metricsgenerationprocessor/processor_test.go:396]: This configuration is now done by the file `testdata/input_metric_types/config.yaml`.

**Review by atoulme — APPROVED** [review]

File: processor/metricsgenerationprocessor/utils.go [hunk 4]
```diff
@@ -65,6 +65,25 @@ func generateMetrics(rm pmetric.ResourceMetrics, operand2 float64, rule internal
 			}
 		}
 	}
+
+	// Remove any calculated metric if there are no data points
```
[the remark this answers is not in the document]
Comment (crobert-1) [reply processor/metricsgenerationprocessor/utils.go]: Good idea, I've implemented it as requested. 👍

**Review by dmitryax — APPROVED** [review]

**Pull request merged**